File: Java/Trapping_Rain_Water.java
package Java;

public class Trapping_Rain_Water {
    public int trap(int[] height) {
        int max_rain = 0;
        int max = height[0];
        int max_left[] = new int[height.length], max_right[] = new int[height.length];
        for (int i = 0; i < height.length; i++) {
            max_left[i] = Math.max(height[i], max);
            max = max_left[i];
        }

        max = height[height.length - 1];

        for (int i = height.length - 1; i >= 0; i--) {
            max_right[i] = Math.max(height[i], max);
            max = max_right[i];
        }

        for (int i = 0; i < height.length; i++) {
            max_rain = max_rain + Math.min(max_left[i], max_right[i]) - height[i];
        }

        return max_rain;
    }
}

File: Python/Trapping_Rain_Water.py
from turtle import left


class Solution(object):
    def trap(self, height):
        l, r = 0, len(height) - 1
        leftmax, rightmax = height[l], height[r]
        ans = 0

        while l < r:
            if leftmax < rightmax:
                l += 1
                leftmax = max(leftmax, height[l])
                ans += leftmax - height[l]
            else:
                r -= 1
                rightmax = max(rightmax, height[r])
                ans += rightmax - height[r]

        return ans